### DISCUSSION TITLE
leave formatting to sqlfmt rather than processing sqlfmt diff output

### DISCUSSION THREAD

**Discussion by ozgenbaris1** [AltimateAI/vscode-dbt-power-user#315] [General]
Hello! 

I have a bad habit of saving files by pressing ctrl + s rapidly. When I do that, formatting behaves weirdly. Some lines are swapped, some of the lines are deleted etc.

I suppose it is because of processing the `sqlfmt --diff` output manually [here](https://github.com/innoverio/vscode-dbt-power-user/blob/5e5010a5150ab1dc15f869eb9bcae476b42a7c0a/src/document_formatting_edit_provider/dbtDocumentFormattingEditProvider.ts#L32-L95). Is there a specific reason for choosing this approach rather than not directly running `sqlfmt <file>` and leaving the formatting job to sqlfmt itself?

For further details, I have a simple sql file as below:

```sql
{{
    config(
        materialized='ephemeral'
    )
}}

select *
from {{ source("raw", 'foo') }}
```

When I run the command `sqlfmt ./foo.sql` here is the formatted file:

```sql
{{ config(materialized="ephemeral") }}

select *
from {{ source("raw", "foo") }}
```

However, if I save the file and leave the formatting to the extension itself. Here is the result:

```sql
{{ config(materialized="ephemeral") }}
from {{ source("raw", "foo") }}

select *
```

I don't know if it's a bug. I'd appreciate your answers.

**Comment by mdesmet** [comment]
Does this only happen if you save very quickly? I think you can just enable autosave in vscode as well, I don't remember having this issue.

Applying the diff is how formatters should work in vscode. See https://code.visualstudio.com/blogs/2016/11/15/formatters-best-practices

**Reply by ozgenbaris1** [reply]
At first sight, I thought it's because of my rapid saves but as I updated my question, it seems the problem isn't related to it. Btw, I'm sharing the output from the command that I guess being run by extension.

```
❯ sqlfmt ./foo.sql --diff --no-progressbar --quiet
1 file failed formatting check.
0 files passed formatting check.
foo.sql failed formatting check.
--- source_query
+++ formatted_query
@@ -1,8 +1,4 @@
-{{
-    config(
-        materialized='ephemeral'
-    )
-}}
+{{ config(materialized="ephemeral") }}
 
 select *
-from {{ source("raw", 'foo') }}
\ No newline at end of file
+from {{ source("raw", "foo") }}
```

**Reply by mdesmet** [reply]
As you were able to locate the code, maybe you can help looking into this and potentially fixing it?

**Reply by ozgenbaris1** [reply]
do you have a guideline for contributing and testing the extension locally?

**Reply by mdesmet** [reply]
Open up the extension in vscode
Make code changes
Launch the extension in debug mode and open your favourite dbt project

<img width="257" alt="image" src="https://user-images.githubusercontent.com/9265503/217817447-11c3afda-1b62-4d4b-8309-e032866b25e6.png">

**Reply by mdesmet** [reply]
I will probably add something more thorough later on in the codebase but hopefully this already gets you started!

**Reply by ozgenbaris1** [reply]
Hey, I opened https://github.com/innoverio/vscode-dbt-power-user/pull/324. Can you please take a look?
